File: Day084_2025/Q084_GFG.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

// User function Template for C++
class Solution {
    bool evaluate(bool b1, bool b2, char op) {
        if (op == '&') return b1 & b2;
        if (op == '|') return b1 | b2;
        return b1 ^ b2;
    }
    
  public:
    int countWays(string &s) {
        // Q. Boolean Parenthesization
        // code here
        int n = s.size();
        
        // dp[i][j][0] -> False count, dp[i][j][1] -> True count
        vector<vector<vector<int>>> dp(n, vector<vector<int>>(n, vector<int>(2, 0)));

        // Initialize base cases
        for (int i = 0; i < n; i += 2) {
            dp[i][i][1] = (s[i] == 'T');
            dp[i][i][0] = (s[i] == 'F');
        }

        // Iterate over increasing lengths of subexpressions
        for (int length = 2; length < n; length += 2) {
            for (int i = 0; i < n - length; i += 2) {
                int j = i + length;
                dp[i][j][0] = dp[i][j][1] = 0;

                for (int k = i + 1; k < j; k += 2) {
                    char op = s[k];

                    int leftTrue = dp[i][k - 1][1];
                    int leftFalse = dp[i][k - 1][0];
                    int rightTrue = dp[k + 1][j][1];
                    int rightFalse = dp[k + 1][j][0];

                    // True evaluations
                    if (evaluate(true, true, op)) dp[i][j][1] += leftTrue * rightTrue;
                    if (evaluate(true, false, op)) dp[i][j][1] += leftTrue * rightFalse;
                    if (evaluate(false, true, op)) dp[i][j][1] += leftFalse * rightTrue;
                    if (evaluate(false, false, op)) dp[i][j][1] += leftFalse * rightFalse;

                    // False evaluations
                    if (!evaluate(true, true, op)) dp[i][j][0] += leftTrue * rightTrue;
                    if (!evaluate(true, false, op)) dp[i][j][0] += leftTrue * rightFalse;
                    if (!evaluate(false, true, op)) dp[i][j][0] += leftFalse * rightTrue;
                    if (!evaluate(false, false, op)) dp[i][j][0] += leftFalse * rightFalse;
                }
            }
        }

        return dp[0][n - 1][1];
    }
};


//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        string s;
        cin >> s;
        Solution ob;
        int ans = ob.countWays(s);
        cout << ans << endl;
        cout << "~" << endl;
    }
}
// } Driver Code Ends

File: Day084_2025/Q084_GFG.py
class Solution:
    def countWays(self, s: str) -> int:
        n = len(s)
        dp = [[[0, 0] for _ in range(n)] for _ in range(n)]

        # Initialize base cases
        for i in range(0, n, 2):
            dp[i][i][1] = 1 if s[i] == 'T' else 0
            dp[i][i][0] = 1 if s[i] == 'F' else 0

        # Iterate over increasing lengths of subexpressions
        for length in range(2, n, 2):
            for i in range(0, n - length, 2):
                j = i + length
                dp[i][j][0] = dp[i][j][1] = 0

                for k in range(i + 1, j, 2):
                    op = s[k]

                    leftTrue, leftFalse = dp[i][k - 1][1], dp[i][k - 1][0]
                    rightTrue, rightFalse = dp[k + 1][j][1], dp[k + 1][j][0]

                    # True evaluations
                    if self.evaluate(True, True, op): dp[i][j][1] += leftTrue * rightTrue
                    if self.evaluate(True, False, op): dp[i][j][1] += leftTrue * rightFalse
                    if self.evaluate(False, True, op): dp[i][j][1] += leftFalse * rightTrue
                    if self.evaluate(False, False, op): dp[i][j][1] += leftFalse * rightFalse

                    # False evaluations
                    if not self.evaluate(True, True, op): dp[i][j][0] += leftTrue * rightTrue
                    if not self.evaluate(True, False, op): dp[i][j][0] += leftTrue * rightFalse
                    if not self.evaluate(False, True, op): dp[i][j][0] += leftFalse * rightTrue
                    if not self.evaluate(False, False, op): dp[i][j][0] += leftFalse * rightFalse

        return dp[0][n - 1][1]

    def evaluate(self, b1: bool, b2: bool, op: str) -> bool:
        if op == '&': return b1 and b2
        if op == '|': return b1 or b2
        return b1 ^ b2
